packages/sdk/client-services/src/packlets/vault/iframe-compatibility-runtime.ts

```diff
@@ -0,0 +1,93 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { Trigger } from '@dxos/async';
+import { Config } from '@dxos/config';
+import { logInfo } from '@dxos/log';
+import { MemorySignalManager, MemorySignalManagerContext, WebsocketSignalManager } from '@dxos/messaging';
+import { createWebRTCTransportFactory, NetworkManager } from '@dxos/network-manager';
+import { createProtoRpcPeer, ProtoRpcPeer, RpcPort } from '@dxos/rpc';
+import { MaybePromise } from '@dxos/util';
+
+import { clientServiceBundle, ClientServices, ClientServicesHost } from '../services';
+
+export type IFrameCompatibilityRuntimeParams = {
+  configProvider: () => MaybePromise<Config>;
+  appPort: RpcPort;
+};
+
+export class IFrameCompatibilityRuntime {
+  private readonly _configProvider: () => MaybePromise<Config>;
+  private readonly _transportFactory = createWebRTCTransportFactory();
+  private readonly _ready = new Trigger<Error | undefined>();
+  private readonly _getServices: (checkContext?: boolean) => Promise<ClientServicesHost>;
+  private readonly _appPort: RpcPort;
+  private _clientServices!: ClientServicesHost;
+  private _clientRpc!: ProtoRpcPeer<ClientServices>;
+  private _config!: Config;
+
+  @logInfo
+  public origin?: string;
+
+  constructor({ configProvider, appPort }: IFrameCompatibilityRuntimeParams) {
+    this._configProvider = configProvider;
+    this._appPort = appPort;
+    this._getServices = async (checkContext = true) => {
+      if (checkContext && !this._clientServices.isOpen) {
+        throw new Error('Service context is closed');
+      }
+
+      const error = await this._ready.wait();
+      if (error !== undefined) {
+        throw error;
+      }
+      return this._clientServices;
+    };
+  }
+
+  async start() {
+    try {
+      this._config = await this._configProvider();
+      const signalServer = this._config.get('runtime.services.signal.server');
+      this._clientServices = new ClientServicesHost({
+        compatibilityMode: true,
+        config: this._config,
+        networkManager: new NetworkManager({
+          log: true,
+          signalManager: signalServer
+            ? new WebsocketSignalManager([signalServer])
+            : new MemorySignalManager(new MemorySignalManagerContext()), // TODO(dmaretskyi): Inject this context.
+          transportFactory: this._transportFactory
+        })
+      });
+
+      this._clientRpc = createProtoRpcPeer({
+        exposed: clientServiceBundle,
+        handlers: {
+          DataService: async () => (await this._getServices()).services.DataService,
```

Comment:
Try something like this to avoid boilerplate?

```ts
const getService = async (cb) => cb(await this._getServices();

DataService: getService(services => services.DataService),
...
```